lib/libp2p/libp2p.go

```diff
@@ -0,0 +1,681 @@
+	if len(cfg) == 0 {
+		log.Info("No libp2p configuration found")
+		return nil, nil
+	}
+
+	ret := make(map[address.Address]*libp2pCfg)
+
+	for _, c := range cfg {
+		maddr, err := address.NewFromString(c.Miner)
+		if err != nil {
+			return nil, err
+		}
+		mid, err := address.IDFromAddress(maddr)
+		if err != nil {
+			return nil, err
+		}
+
+		ret[maddr] = &libp2pCfg{
+			id: int64(mid),
+		}
+
+		var lstnAddrStr, annAddrStr, naAddrStr []string
+
+		for _, l := range c.ListenAddresses {
+			listenAddr, err := multiaddr.NewMultiaddr(l)
+			if err != nil {
+				return nil, xerrors.Errorf("parsing listen address: %w", err)
+			}
+			ret[maddr].ListenAddr = append(ret[maddr].ListenAddr, listenAddr)
+			lstnAddrStr = append(lstnAddrStr, listenAddr.String())
+		}
+
+		for _, a := range c.AnnounceAddresses {
+			announceAddr, err := multiaddr.NewMultiaddr(a)
+			if err != nil {
+				return nil, xerrors.Errorf("parsing announce address: %w", err)
+			}
+			ret[maddr].AnnounceAddr = append(ret[maddr].AnnounceAddr, announceAddr)
+			annAddrStr = append(annAddrStr, announceAddr.String())
+		}
+
+		for _, na := range c.NoAnnounceAddresses {
+			noAnnounceAddr, err := multiaddr.NewMultiaddr(na)
+			if err != nil {
+				return nil, xerrors.Errorf("parsing no announce address: %w", err)
+			}
+			ret[maddr].NoAnnounceAddr = append(ret[maddr].NoAnnounceAddr, noAnnounceAddr)
+			naAddrStr = append(naAddrStr, noAnnounceAddr.String())
+		}
+
+		var privKey []byte
+
+		err = db.QueryRow(ctx, `SELECT insert_or_update_libp2p ($1, $2, $3, $4, $5)`, int64(mid),
+			strings.Join(lstnAddrStr, ","), strings.Join(annAddrStr, ","), strings.Join(naAddrStr, ","), machine).Scan(&privKey)
+		if err != nil {
+			return nil, xerrors.Errorf("acquiring libp2p locks from DB: %w", err)
+		}
+
+		p, err := crypto.UnmarshalPrivateKey(privKey)
+		if err != nil {
+			return nil, xerrors.Errorf("unmarshaling private key: %w", err)
+		}
+
+		ret[maddr].priv = p
+
+	}
+
+	return ret, nil
+}
+
+func MakeAddrsFactory(announceAddrs, noAnnounce []multiaddr.Multiaddr) (basichost.AddrsFactory, error) {
+	filters := multiaddr.NewFilters()
+	noAnnAddrs := map[string]bool{}
+	for _, addr := range noAnnounce {
+		f, err := mamask.NewMask(addr.String())
+		if err == nil {
+			filters.AddFilter(*f, multiaddr.ActionDeny)
+			continue
+		}
+		noAnnAddrs[string(addr.Bytes())] = true
+	}
+
+	return func(allAddrs []multiaddr.Multiaddr) []multiaddr.Multiaddr {
+		var addrs []multiaddr.Multiaddr
+		if len(announceAddrs) > 0 {
+			addrs = announceAddrs
+		} else {
+			addrs = allAddrs
+		}
+
+		var out []multiaddr.Multiaddr
+		for _, maddr := range addrs {
+			// check for exact matches
+			ok := noAnnAddrs[string(maddr.Bytes())]
+			// check for /ipcidr matches
+			if !ok && !filters.AddrBlocked(maddr) {
+				out = append(out, maddr)
+			}
+		}
+		return out
+	}, nil
+}
+
+var netlog = logging.Logger("mk12-net")
```

Comment:
Moved libp2p under market. Libp2p is only for mk12 really. At first, those were in mk12 then I thought since libp2p only has one job, a clear separation would be nice. This would allow us to attach mk12 to HTTP in future if we want.